File: Doc-Scanner-X/Doc-Scanner-X/WebViewController.m
//
//  WebViewController.m
//  Doc-Scanner-X
//
//  Created by dynamsoft on 2017/11/27.
//  Copyright © 2017年 com.dynamsoft. All rights reserved.
//

#import "WebViewController.h"

@interface WebViewController ()

@end

@implementation WebViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    _webView = [[WKWebView alloc] initWithFrame:CGRectZero configuration:[WKWebViewConfiguration new]];
    _webView.navigationDelegate = self;
    _webView.allowsBackForwardNavigationGestures = YES;
    NSURL* myURL = [[NSURL alloc] initWithString:@"https://developer.dynamsoft.com/dws/ios-guide"];
    NSURLRequest* myRequest = [[NSURLRequest alloc] initWithURL:myURL];
    [_webView loadRequest:myRequest];
    _webView.frame = CGRectMake(0, 64, [UIScreen mainScreen].bounds.size.width, [UIScreen mainScreen].bounds.size.height-64);
    
    _navigationBar = [[CustomNavigationBar alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, 64)];
    self.navigationItem.title = @"Developer Center";
    [_navigationBar pushNavigationItem:self.navigationItem animated:YES];
    MainViewController* parentVC = (MainViewController*)self.parentViewController;
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[[UIImage imageNamed:@"icon_back"] imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal] style:UIBarButtonItemStylePlain target:parentVC action:@selector(backToHome)];
    
    _refreshControl = [UIRefreshControl new];
    _refreshControl.attributedTitle = [[NSAttributedString alloc] initWithString:@"Pull to refresh"];
    [_refreshControl addTarget:self action:@selector(refreshWeb) forControlEvents:UIControlEventValueChanged];
    _webView.scrollView.bounces = YES;
    [_webView.scrollView addSubview:_refreshControl];
    
    self.view.frame = CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, [UIScreen mainScreen].bounds.size.height);
    [self.view addSubview:_navigationBar];
    [self.view addSubview:_webView];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [_webView addObserver:self forKeyPath:@"canGoBack" options:NSKeyValueObservingOptionNew context:nil];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [_webView removeObserver:self forKeyPath:@"canGoBack"];
}

- (void)refreshWeb {
    [_refreshControl endRefreshing];
    [_webView reload];
}

- (void)back {
    [_webView goBack];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if (_webView.canGoBack) {
        self.navigationItem.leftBarButtonItem.target = self;
        self.navigationItem.leftBarButtonItem.action = @selector(back);
    } else {
        MainViewController* parentVC = (MainViewController*)self.parentViewController;
        self.navigationItem.leftBarButtonItem.target = parentVC;
        self.navigationItem.leftBarButtonItem.action = @selector(backToHome);
    }
}

- (void)webView:(WKWebView *)webView didStartProvisionalNavigation:(WKNavigation *)navigation {
    _activityIndicator.backgroundColor = [UIColor clearColor];
    _activityIndicator.center = CGPointMake([UIScreen mainScreen].bounds.size.width/2, [UIScreen mainScreen].bounds.size.height/2);
    _activityIndicator.hidesWhenStopped = YES;
    [self.view addSubview:_activityIndicator];
    [_activityIndicator startAnimating];
}

- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation {
    [_activityIndicator stopAnimating];
    [webView evaluateJavaScript:@"document.getElementById('liveChatMobile').style.visibility='hidden'" completionHandler:nil];
}

- (void)webView:(WKWebView *)webView didFailNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    [_activityIndicator stopAnimating];
}

- (void)webView:(WKWebView *)webView didFailProvisionalNavigation:(WKNavigation *)navigation withError:(NSError *)error {
    [_activityIndicator startAnimating];
    switch (error.code) {
        case NSURLErrorTimedOut:
            printf("timeout");
            break;
        case NSURLErrorNotConnectedToInternet:
            printf("no Internet");
            break;
        default:
            printf("other error");
            break;
    }
}


@end

File: DocScannerX/DocScannerX/WebViewController.swift
//
//  WebViewController.swift
//  DCS SuperScanner
//
//  Created by dynamsoft on 2017/11/1.
//  Copyright © 2017年 com.dynamsoft. All rights reserved.
//

import UIKit
import WebKit
import JavaScriptCore


class WebViewController: UIViewController, WKNavigationDelegate {
    // MARK: - Properties
    
    var webView: WKWebView!
    var navigationBar: UINavigationBar!
    let fullScreenSize = UIScreen.main.bounds.size
    let progressIndicator = UIActivityIndicatorView(activityIndicatorStyle: .gray)
    var refreshControl: UIRefreshControl?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        /// Initialize `webView`.
        let webConfiguration = WKWebViewConfiguration()
        webView = WKWebView(frame: .zero, configuration: webConfiguration)
        webView.navigationDelegate = self
        let myURL = URL(string: "https://developer.dynamsoft.com/dws/ios-guide")
        let myRequest = URLRequest(url: myURL!)
        webView.load(myRequest)
        webView.frame = CGRect(x: 0, y: 64, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.height-64)
        webView.allowsBackForwardNavigationGestures = true
        
        /// Initialize `navigationBar`.
        navigationBar = CustomNavigationBar(frame: CGRect(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: 64))
        navigationItem.title = "Developer Center"
        navigationBar.pushItem(self.navigationItem, animated: true)
        let parentVC = self.parent as? MainViewController
        navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "icon_back")?.withRenderingMode(UIImageRenderingMode.alwaysOriginal), style: .plain, target: parentVC, action: #selector(parentVC?.backToHome))
        
        /// Initialize `refreshControl`.
        refreshControl = UIRefreshControl()
        refreshControl?.attributedTitle = NSAttributedString(string: "Pull to refresh")
        refreshControl?.addTarget(self, action: #selector(refreshWeb), for: .valueChanged)
        webView.scrollView.addSubview(refreshControl!)
        
        view.frame = CGRect(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.height)
        
        /// Add `navigationBar` and `webView` as subviews of `view`.
        view.addSubview(navigationBar)
        view.addSubview(webView)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        webView.addObserver(self, forKeyPath: "canGoBack", options: .new, context: nil)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        webView.removeObserver(self, forKeyPath: "canGoBack")
    }
    
    func refreshWeb() {
        webView.reload()
        refreshControl?.endRefreshing()
    }
    
    func back(){
        webView.goBack()
    }
    
    // MARK: - Observer function
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if webView.canGoBack {
            navigationItem.leftBarButtonItem?.target = self
            navigationItem.leftBarButtonItem?.action = #selector(back)
        } else {
            let parentVC = parent as? MainViewController
            navigationItem.leftBarButtonItem?.target = parentVC
            navigationItem.leftBarButtonItem?.action = #selector(parentVC?.backToHome)
        }
    }
    
    // MARK: - WKNavigationDelegate
    
    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        progressIndicator.backgroundColor = UIColor.clear
        progressIndicator.center = CGPoint(x: fullScreenSize.width/2, y: fullScreenSize.height/2)
        progressIndicator.hidesWhenStopped = true
        view.addSubview(progressIndicator)
        progressIndicator.startAnimating()
        
    }
    
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        progressIndicator.stopAnimating()
        webView.evaluateJavaScript("document.getElementById('liveChatMobile').style.visibility='hidden'", completionHandler: nil)
    }
    
    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        progressIndicator.stopAnimating()
    }

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        progressIndicator.stopAnimating()
        if let err = error as? URLError {
            switch err.code {
            case URLError.timedOut:
                print("timeout")
            case URLError.notConnectedToInternet:
                print("no Internet")
            default:
                print("other error")
            }
        }
    }

}
